Use Link design colors on Link in MPE

carlosmuvi-stripe opened the pull request:
# Summary
Uses Link Design v2 colors on Link in MPE. See this template. See
- [This reference](https://www.figma.com/design/OGzRm5DunZ6ih4qHo73vEO/%F0%9F%8D%8F%F0%9F%93%B1-Mobile-Payment-Element?node-id=488-4329&t=wzYfK3AEcK1hO6h6-4) for Link in MPE specifically
- [General reference of Link colors](https://www.figma.com/design/mhgvqbODPxfR4c0j2u3ySn/%E2%9D%87%EF%B8%8F-Link-Design-System-2.0?node-id=14875-83637&t=KONmgsfYCMU7rQYY-4).

Additional changes / worth mentioning:

- Replaced system spinner by [Link spinner](https://www.figma.com/design/mhgvqbODPxfR4c0j2u3ySn/%E2%9D%87%EF%B8%8F-Link-Design-System-2.0?node-id=8511-29258&t=iAvQllUpLMMDNTfR-4), that plays well with the new color definition
- Used [Link error banners](https://www.figma.com/design/mhgvqbODPxfR4c0j2u3ySn/%E2%9D%87%EF%B8%8F-Link-Design-System-2.0?node-id=16236-4818&t=iAvQllUpLMMDNTfR-4)
- Kept primary CTA green. This differs a bit from the Link design system, where primary colors are black (or white in darkmode). 

I'd recommend checking out the branch and testing it locally, plus comparing the screenshot diff directly on the IDE.

# Motivation
https://jira.corp.stripe.com/browse/LINK_MOBILE-98

# Testing
<!-- How was the code tested? Be as specific as possible. -->
- [ ] Added tests
- [X] Modified tests (screenshot tests)
- [X] Manually verified

# Changelog
<!-- Is this a notable change that affects users? If so, add a line to `CHANGELOG.md` and prefix the line with one of the following:
    - [Added] for new features.
    - [Changed] for changes in existing functionality.
    - [Deprecated] for soon-to-be removed features.
    - [Removed] for now removed features.
    - [Fixed] for any bug fixes.
    - [Security] in case of vulnerabilities.
-->

carlosmuvi-stripe commented on paymentsheet/res/drawable/stripe_link_error.xml:
updated to use the new sail error icon.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/theme/Color.kt:
```diff
@@ -1,3 +1,5 @@
+@file:Suppress("unused")
```
I'll revert this once the palette is finalized, planning on keeping the colors used in the palette and removing the rest.

carlosmuvi-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui.inline_LinkFieldsScreenshotTest_testErrorMessage[DarkTheme,DefaultFont].png:
don't trust darkmode colors just yet.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/paymentmenthod/PaymentMethodScreen.kt:
```diff
@@ -55,10 +55,13 @@ internal fun PaymentMethodBody(
         )
 
-        StripeThemeForLink {
+        StripeThemeForLink(
+            componentBorder = LinkTheme.colors.surfacePrimary,
+            componentDivider = LinkTheme.colors.surfacePrimary
+        ) {
```
This forces separators on filled containers (surface_secondary) to look "transparent" (matching the overall background, surface_primary).

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/theme/Color.kt:
```diff
@@ -72,62 +59,104 @@ internal object LinkThemeConfig {
+        iconCritical = Critical500
     )
+
+    /**
```
Instead of exposing colors directly across Views, unifying non pure Link Design systems workarounds here. That should help iterating in the future as we get closer to pure Link designs.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/inline/LinkInlineSignupFields.kt:
```diff
@@ -160,4 +177,56 @@ internal fun LinkInlineSignupFields(
     }
 }
 
+@Composable
+private fun ErrorText(
+    text: String,
+    modifier: Modifier = Modifier,
+) {
+    Row(
+        modifier = modifier,
+        verticalAlignment = Alignment.CenterVertically
+    ) {
+        Icon(
+            modifier = Modifier
+                .padding(horizontal = 10.dp, vertical = 12.dp)
+                .size(20.dp),
+            painter = painterResource(id = R.drawable.stripe_ic_sail_warning_circle),
+            contentDescription = null,
+            tint = MaterialTheme.colors.error
+        )
+        HyperlinkedText(
+            text = text,
+            modifier = Modifier
+                .padding(vertical = 12.dp),
+            color = MaterialTheme.colors.error,
+            style = TextStyle(
+                fontFamily = FontFamily.Default,
+                fontWeight = FontWeight.Normal,
+                fontSize = 14.sp,
+                lineHeight = 20.sp
+            )
+        )
+    }
+}
```
separated Inline signup error view from the Link one, so that we can update the Link flow designs without impacting  the inline view.

carlosmuvi-stripe commented on paymentsheet/src/test/java/com/stripe/android/link/ui/wallet/WalletScreenTest.kt:
```diff
@@ -328,52 +328,6 @@ internal class WalletScreenTest {
-
-        composeTestRule.waitForIdle()
-
-        onWalletDialogTag()
```
The dialog shows up but doesn't seem to be rendered on the compose tree 😢 

<img width="466" alt="image" src="https://github.com/user-attachments/assets/2155072a-e477-44e6-aec1-c106755a0a6f" />

carlosmuvi-stripe commented on paymentsheet/res/drawable/stripe_link_add_green.xml:
Per design recommendation, we've removed the add payment method icon.

carlosmuvi-stripe commented on paymentsheet/res/drawable/stripe_link_error.xml:
Updated the error icon to use the sail reference.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/theme/Color.kt:
```diff
@@ -72,62 +59,104 @@ internal object LinkThemeConfig {
 
 @Composable
 internal fun StripeThemeForLink(
+    materialPrimary: Color = LinkTheme.colors.textPrimary,
```
For certain material-based Stripe components, we want to override the primary color to maintain consistency and match designs.

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/ErrorText.kt:
```diff
@@ -71,28 +85,21 @@ internal fun ErrorText(
-                style = style.textStyle
-            )
-        }
+    Row(
```
We're not using the `ErrorText` view on inline signup (since we don't want the redesign to apply there for now).

lng-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/wallet/WalletScreen.kt:
```diff
@@ -563,16 +554,11 @@ private fun AddPaymentMethodRow(
+        Spacer(modifier = Modifier.width(HorizontalPadding))
         Text(
             text = stringResource(R.string.stripe_add_payment_method),
             modifier = Modifier.padding(end = HorizontalPadding),
```
Can we do this instead of adding the `Spacer` above?
```suggestion
            modifier = Modifier.padding(horizontal = HorizontalPadding),
```

lng-stripe commented on paymentsheet/src/test/java/com/stripe/android/link/ui/LinkScreenshotTheme.kt:
```diff
@@ -0,0 +1,18 @@
+import com.stripe.android.link.theme.LinkTheme
+
+@Composable
+internal fun LinkScreenshotTheme(
```
nit: rename to `LinkScreenshotSurface` to make it clear it's adding a background
```suggestion
internal fun LinkScreenshotSurface(
```

lng-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui.updatecard_UpdateCardScreenshotTest_testScreen[Error][DarkTheme,DefaultFont].png:
I think the label color in the "Card information" section is off: should be brighter to match those under "Billing address"

<img width="426" alt="Screenshot 2025-05-13 at 3 13 12 PM" src="https://github.com/user-attachments/assets/5312e3c8-b221-474c-8761-c3466a8e40c7" />

carlosmuvi-stripe replied:
good catch - the update card component was setting all labels to be colored as disabled via alpha. Removed that so that `textFieldColors.label` theming (that takes into account disabled / enabled) is used. 

With that we can clearly see the fields that are disabled (CC number and CVC) 

<img width="796" alt="image" src="https://github.com/user-attachments/assets/8c5e7e1d-72fe-4fe3-8897-e2a8ea88ae66" />

carlosmuvi-stripe replied:
@lng-stripe on top of the above we updated the text tertiary on dark mode, that now is lighter and more visible!

lng-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui.updatecard_UpdateCardScreenshotTest_testScreen[Error][LightTheme,DefaultFont].png:
Same but opposite issue as dark mode label colors. See other comment

lng-stripe reviewed (dismissed):
It's looking so good! Just spotted some minor things here and there

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/verification/VerificationScreen.kt:
```diff
@@ -246,13 +270,14 @@ private fun ChangeEmailRow(
 
+@OptIn(ExperimentalComposeUiApi::class)
 @Composable
 private fun ResendCodeButton(
```
this view is changing in a follow up PR redesigning the verification dialog, that's why I kept it as is!

carlosmuvi-stripe commented:
@lng-stripe addressed all feedback in [this commit](https://github.com/stripe/stripe-android/pull/10782/commits/ca44436e0eac1c6f87d225ce2ab1d7f33a6ad86d)! Highlighting:

- centered errors
- use non-disabled label when TextField is not disabled

tillh-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui.wallet_WalletScreenScreenshotTest_testExpandedStateWithUnavailable[Portrait,LightTheme].png:
The bank icon background isn't really visible anymore. Should we make it distinct from the picker background?

carlosmuvi-stripe replied:
updated it to match the default badge definition. Also noticed the tertiary color is now lighter on dark mode, updated that as well: https://github.com/stripe/stripe-android/pull/10782/commits/34ce157bdc1b5ecea0e768d76069cc7a228fecc7

tillh-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.paymentsheet.verticalmode_NewPaymentMethodRowButtonScreenshotTest_testUnselectedState[DarkTheme,CrazyAppearance,LargeFont].png:
Why are MPE elements impacted by this change?

carlosmuvi-stripe replied:
Can't spot the change when comparing screenshots, but they seem to fail if not regenerated 🤔

tillh-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui_SecondaryButtonScreenshotTest_testDisabledState[DarkTheme,DefaultFont].png:
Are we changing the secondary buttons to plain buttons (without background) in a later pull request?

carlosmuvi-stripe replied:
we'll for now use branded and secondary as defined in Figma. 
<img width="338" alt="image" src="https://github.com/user-attachments/assets/6416ae6b-5017-4cbe-9cb4-61506e428372" />

carlosmuvi-stripe commented on paymentsheet/src/test/snapshots/images/com.stripe.android.link.ui.wallet_PaymentDetailsListItemScreenShotTest_testCardEnabledAndUnavailable[LightTheme,DefaultFont].png:
this is intentional - see definition here, and distinction between `iconCritical` and `textCritical` - https://www.figma.com/design/OGzRm5DunZ6ih4qHo73vEO/%F0%9F%8D%8F%F0%9F%93%B1-Mobile-Payment-Element?node-id=603-6206&m=dev

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/PrimaryButton.kt:
```diff
@@ -60,19 +63,18 @@ internal fun PrimaryButton(
+                            .size(20.dp)
+                            .semantics { testTag = ProgressIndicatorTestTag },
+                        backgroundColor = LinkTheme.colors.surfaceBackdrop.copy(alpha = 0.1f),
+                        strokeWidth = 4.dp,
```
we can likely reduce it a bit! width is not specified [in Figma](https://www.figma.com/design/mhgvqbODPxfR4c0j2u3ySn/%E2%9D%87%EF%B8%8F-Link-Design-System-2.0?node-id=15148-140088&m=dev).

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/signup/SignUpScreen.kt:
```diff
@@ -193,11 +194,17 @@ private fun SecondaryFields(
     onSignUpClick: () -> Unit
 ) {
+    var emailFocused by remember { mutableStateOf(false) }
+    var nameFocused by remember { mutableStateOf(false) }
```
https://github.com/stripe/stripe-android/pull/10782/commits/6ddeedce6508058d3392a06c7d5fe2edbc0773d3

carlosmuvi-stripe commented on paymentsheet/src/main/java/com/stripe/android/link/ui/signup/SignUpScreen.kt:
```diff
@@ -143,6 +145,7 @@ private fun EmailCollectionSection(
     signUpScreenState: SignUpScreenState,
     focusRequester: FocusRequester = remember { FocusRequester() }
 ) {
+    var focused by remember { mutableStateOf(false) }
```
https://github.com/stripe/stripe-android/pull/10782/commits/6ddeedce6508058d3392a06c7d5fe2edbc0773d3
